api/metric/api.go

```diff
@@ -16,66 +16,97 @@ package metric
 
 import (
 	"context"
+	"time"
 
 	"go.opentelemetry.io/api/core"
 	"go.opentelemetry.io/api/registry"
 	"go.opentelemetry.io/api/unit"
 )
 
-type MetricType int
+// Kind categorizes different kinds of metric.
+type Kind int
 
+//go:generate stringer -type=Kind
 const (
-	Invalid    MetricType = iota
-	Gauge                 // Supports Set()
-	Cumulative            // Supports Inc()
+	Invalid        Kind = iota
+	CumulativeKind      // Supports Add()
+	GaugeKind           // Supports Set()
+	MeasureKind         // Supports Record()
 )
 
-type Meter interface {
-	// TODO more Metric types
-	GetFloat64Gauge(ctx context.Context, gauge *Float64GaugeHandle, labels ...core.KeyValue) Float64Gauge
+// Recorder is the implementation-level interface to Set/Add/Record individual metrics.
+type Recorder interface {
+	// RecordAt allows the SDK to observe a single metric event
+	RecordAt(ctx context.Context, t time.Time, value float64)
 }
 
-type Float64Gauge interface {
-	Set(ctx context.Context, value float64, labels ...core.KeyValue)
+// Meter is an interface to the metrics portion of the OpenTelemetry SDK.
+type Meter interface {
+	// RecorderFor returns a handle for observing single measurements.
+	RecorderFor(context.Context, Instrument, ...core.KeyValue) Recorder
+
+	// RecordBatch atomically records a batch of measurements.
+	RecordBatch(context.Context, ...Measurement)
```

Comment:
atomically may be too strict a requirement.  What if the buffer can only fit 2 of them?  I think it's fine to accept those two and trigger back pressure logic for the rest.

Reply:
When this was discussed, it was being done explicitly to support recording inter-dependent metrics, where you wouldn't want an update to be split apart because it could create inconsistency. 

I'm not sure the caller ever wants us to block. I'm starting to think this API should return an error in this sort of condition,.

Reply:
If you return errors you're asking the caller to check them, which makes metric reporting verbose.  It may be better to register an "onerror" callback with the system that takes measurements that fail, and the library user can decide what to do.